pkg/pipeline/input/builder/audio.go

```diff
@@ -0,0 +1,299 @@
+
+	switch {
+	case strings.EqualFold(codec.MimeType, string(params.MimeTypeOpus)):
+		if err := src.Element.SetProperty("caps", gst.NewCapsFromString(
+			fmt.Sprintf(
+				"application/x-rtp,media=audio,payload=%d,encoding-name=OPUS,clock-rate=%d",
+				codec.PayloadType, codec.ClockRate,
+			),
+		)); err != nil {
+			return err
+		}
+
+		rtpOpusDepay, err := gst.NewElement("rtpopusdepay")
+		if err != nil {
+			return err
+		}
+
+		opusDec, err := gst.NewElement("opusdec")
+		if err != nil {
+			return err
+		}
+		if err = opusDec.SetProperty("use-inband-fec", true); err != nil {
+			return err
+		}
+
+		a.decoder = append(a.decoder, rtpOpusDepay, opusDec)
+
+	default:
+		return errors.ErrNotSupported(codec.MimeType)
+	}
+
+	audioQueue, err := buildQueue()
+	if err != nil {
+		return err
+	}
+
+	audioConvert, err := gst.NewElement("audioconvert")
+	if err != nil {
+		return err
+	}
+
+	// TODO: sinc-filter-mode=full will use more memory but much less CPU
+	audioResample, err := gst.NewElement("audioresample")
+	if err != nil {
+		return err
+	}
+
+	capsFilter, err := getCapsFilter(p)
+	if err != nil {
+		return err
+	}
+
+	a.decoder = append(a.decoder, audioQueue, audioConvert, audioResample, capsFilter)
+	return nil
+}
+
+func (a *AudioInput) buildMixer(p *params.Params) error {
+	audioTestSrc, err := gst.NewElement("audiotestsrc")
+	if err != nil {
+		return err
+	}
+	if err = audioTestSrc.SetProperty("volume", 0.0); err != nil {
+		return err
+	}
+	if err = audioTestSrc.SetProperty("do-timestamp", true); err != nil {
+		return err
+	}
+	if err = audioTestSrc.SetProperty("is-live", true); err != nil {
+		return err
+	}
+	audioCaps, err := getCapsFilter(p)
+	if err != nil {
+		return err
+	}
+	a.testSrc = []*gst.Element{audioTestSrc, audioCaps}
+
+	audioMixer, err := gst.NewElement("audiomixer")
+	if err != nil {
+		return err
+	}
+	// set latency slightly higher than max audio appsrc latency
+	if err = audioMixer.SetProperty("latency", uint64(41e8)); err != nil {
+		logger.Errorw("latency", err)
+		return err
+	}
+	capsFilter, err := getCapsFilter(p)
+	if err != nil {
+		return err
+	}
+	a.mixer = []*gst.Element{audioMixer, capsFilter}
+
+	return nil
+}
+
+func (a *AudioInput) buildEncoder(p *params.Params) error {
+	switch p.AudioCodec {
+	case params.MimeTypeOpus:
+		encoder, err := gst.NewElement("opusenc")
+		if err != nil {
+			return err
+		}
+		if err = encoder.SetProperty("bitrate", int(p.AudioBitrate*1000)); err != nil {
+			return err
+		}
+		a.encoder = encoder
+
+	case params.MimeTypeAAC:
+		encoder, err := gst.NewElement("faac")
+		if err != nil {
+			return err
+		}
+		if err = encoder.SetProperty("bitrate", int(p.AudioBitrate*1000)); err != nil {
+			return err
+		}
+		a.encoder = encoder
+
+	default:
+		return errors.ErrNotSupported(string(p.AudioCodec))
+	}
+
+	return nil
+}
+
+func getCapsFilter(p *params.Params) (*gst.Element, error) {
+	var caps *gst.Caps
+	switch p.AudioCodec {
+	case params.MimeTypeOpus:
```

Comment:
should we also set websocket / raw output to two channel and 48k for consistency for real-time workflows?

Reply:
it's already doing that, audio codec is opus in that case with output type raw